### PR TITLE
Add behaviour pipeline for logging in handlers

### PR DESCRIPTION
**Pull request by Victor-van-Duijnen**
- Add a generic logging behaviour for query and command handlers.
- Divide commands/queries for fetching and creating user inside cfa
-  Add integrations tests for user fetching/creation

### CODE REVIEW

File: Smart.FA.Catalog.IntegrationTests/Smart.FA.Catalog.IntegrationTests.csproj [hunk 1]
```diff
@@ -12,6 +12,7 @@
     <PackageReference Include="Microsoft.EntityFrameworkCore" Version="6.0.1" />
     <PackageReference Include="Microsoft.EntityFrameworkCore.SqlServer" Version="6.0.1" />
     <PackageReference Include="Microsoft.NET.Test.Sdk" Version="17.0.0" />
+    <PackageReference Include="Moq" Version="4.16.1" />
```
Comment (LievenPraet) [inline Smart.FA.Catalog.IntegrationTests/Smart.FA.Catalog.IntegrationTests.csproj:15]: Can we use `NSubstitute` instead of `Moq`?
We want to stay consistent with the packages we use throughout projects.

File: Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs [hunk 2]
```diff
@@ -49,7 +51,19 @@ public async Task CanChangeTrainerName(string firstName, string lastName)
...
+        var trainer = _trainerFactory.CreateFromUser(user);
 
+        context.Trainers.Add(trainer);
+        await context.SaveChangesAsync();
 
-
+        trainer.Should().NotBeNull();
+        trainer!.Name.FirstName.Should().NotBeEmpty();
+        trainer.Name.LastName.Should().NotBeEmpty();
+        trainer.DefaultLanguage.Value.Should().NotBeEmpty();
```
Comment (LievenPraet) [inline Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs:67]: Instead of checking for not empty, you can check the actual values since you have them for your user object

File: Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs [hunk 3]
```diff
@@ -14,6 +15,7 @@ namespace Smart.FA.Catalog.IntegrationTests.TrainingContext;
 public class TrainerTests : IntegrationTestBase
 {
     private readonly TrainerFactory _trainerFactory = new();
+    private readonly UserFactory _userFactory = new();
```
Comment (LievenPraet) [inline Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs:18]: You can use `NSubstitute` to create these instances for you.
This will help in the future in case extra dependencies are introduced in these classes because `NSubstitute` will handle them for you.

File: Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs [hunk 4]
```diff
@@ -14,6 +15,7 @@ namespace Smart.FA.Catalog.IntegrationTests.TrainingContext;
...
 
     [Theory]
     [InlineData("Victor", "vD")]
```
Comment (LievenPraet) [inline Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs:21]: You can add additional tests to test what happens when you pass in empty or invalid values.

File: Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs [hunk 5]
```diff
@@ -49,7 +51,19 @@ public async Task CanChangeTrainerName(string firstName, string lastName)
...
+        await using var context = GivenTrainingContext();
+        var user = _userFactory.CreateClean();
+        var trainer = _trainerFactory.CreateFromUser(user);
 
+        context.Trainers.Add(trainer);
+        await context.SaveChangesAsync();
 
-
+        trainer.Should().NotBeNull();
+        trainer!.Name.FirstName.Should().NotBeEmpty();
```
Comment (LievenPraet) [inline Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs:65]: Any reason you use `!` for some lines but not other lines?

File: Smart.FA.Catalog.Tests.Common/UserFactory.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿using Core.Domain.Dto;
+using Core.Domain.Enumerations;
+
+namespace Smart.FA.Catalog.Tests.Common;
+
+public class UserFactory
+{
+    public UserDto CreateClean()
+    {
+        return new UserDto("1", "John", "Doe", ApplicationType.Account.Name);
```
Comment (LievenPraet) [inline Smart.FA.Catalog.Tests.Common/UserFactory.cs:10]: You could use `Fixture` here as well

**Review by LievenPraet — CHANGES_REQUESTED** [review]

File: Smart.FA.Catalog.IntegrationTests/Smart.FA.Catalog.IntegrationTests.csproj (hunk 1 above)
Comment (Victor-van-Duijnen) [reply Smart.FA.Catalog.IntegrationTests/Smart.FA.Catalog.IntegrationTests.csproj:15]: Alright I can do that, I will have to substitute it in the unit test projects, too so I will do a separate PR for it

File: Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs (hunk 5 above)
Comment (Victor-van-Duijnen) [reply Smart.FA.Catalog.IntegrationTests/TrainingContext/Trainer_Tests.cs:65]: If the trainer is null, it should not pass line #64  in the assert section of the code.  So at line #65 I am absolutely certain the trainer is not null (before it, I can't be certain), so I just make it explicit to the compiler since the IDE would complain that if the trainer is null, he would not be able to access any of its properties.

**Comment by maxime-poulain** [comment]
Please don't create a PR with three different features.
This should have been spitted into three PR.
Or at least don't give it 
>Add behaviour pipeline for logging in handlers
as title.
